### DISCUSSION TITLE
ldap auth

### DISCUSSION THREAD

**Discussion by alxcohen** [DeviceFarmer/stf#803] [Q&A]
Hi there.
Couple questions.
1)RethinkDB looks like not maintained anymore. Is that possible to use mongo etc ?
2) For now i'm using rethinkdb in docker. Stf using locally (not in docker). 
Trying to make a stf as service, execstart = stf local --public-ip **** , but it crushes with this log 
`Aug 12 18:01:12 mob-farm stf[78308]: Module did not self-register: '/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/node_modules/zeromq/build/Release/zmq.node'.
Aug 12 18:01:12 mob-farm stf[78286]: 2024-08-12T15:01:12.294Z FTL/cli:local 78286 [*] Child process had an error ExitError: Exit code "1"
Aug 12 18:01:12 mob-farm stf[78286]:     at ChildProcess.<anonymous> (/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/lib/util/procutil.js:49:23)
Aug 12 18:01:12 mob-farm stf[78286]:     at ChildProcess.emit (node:events:517:28)
Aug 12 18:01:12 mob-farm stf[78286]:     at ChildProcess._handle.onexit (node:internal/child_process:292:12)
Aug 12 18:01:12 mob-farm stf[78286]: 2024-08-12T15:01:12.295Z INF/cli:local 78286 [*] Shutting down all child processes`

Anyways for testing i'm starting app in tmux and everything look great. But now I tried to enable ldap and can't actually understand how. Trying to start like this
`stf auth-ldap --app-url http://ip:7100 \
              --ldap-url ldap://my.domain.com \
              --ldap-search-dn "DC=my,DC=domain,DC=com" \
              --ldap-bind-dn "cn=user,dc=my,dc=domain,dc=com" \
              --ldap-bind-credentials "pass" \
              --ldap-search-filter "(&(memberOf=CN=group,OU=ou,OU=ou,DC=my,DC=domain,DC=com)(sAMAccountName={{username}}))"`

And had error that port is using. Which is means that instance already started with auth-ldap I guess. But how to configure than ?
Thanks.

**Comment by denis99999** [comment]
@alxcohen , regarding your ldap problem, you are using the default port (7100) on which STF listens, which STF service did you launch before the ldap one? Please provide your entire configuration at runtime!

**Reply by alxcohen** [reply]
Please provide your entire configuration at runtime!

using this for launch `stf local --public-ip ****`

**Comment by denis99999** [comment]
@alxcohen , so this is the reason of the port conflict, if you launch `stf local ...` you have to provide the ldap arguments as Options, take a look at the STF help (i.e. `stf local --help`), in your case it should look something like that:

```
  stf local \
    --auth-type ldap \
    --auth-options '["--ldap-url","ldap://my.domain.com","--ldap-search-dn","dc=my,dc=domain,dc=com","--ldap-bind-dn","cn=user,dc=my,dc=domain,dc=com","--ldap-bind-credentials","pass",...' \
    --public-ip YOUR_STF_PUBLIC_IP
    ...

```

**Reply by alxcohen** [reply]
Thanks a lot, it working now, at least starting and asking for ldap, but there is some problems with filter, so can't login (
`2024-08-13T12:53:06.887Z ERR/auth-ldap 97598 [::ffff:127.0.0.1] Unexpected error InvalidCredentialsError: 80090308: LdapErr: DSID-0C090449, comment: AcceptSecurityContext error, data 52e, v3839
    at messageCallback (/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/node_modules/ldapjs/lib/client/client.js:1419:45)
    at Parser.onMessage (/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/node_modules/ldapjs/lib/client/client.js:1089:14)
    at Parser.emit (node:events:519:28)
    at Parser.write (/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/node_modules/ldapjs/lib/messages/parser.js:111:8)
    at Socket.onData (/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/node_modules/ldapjs/lib/client/client.js:1076:22)
    at Socket.emit (node:events:519:28)
    at addChunk (node:internal/streams/readable:559:12)
    at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
    at Readable.push (node:internal/streams/readable:390:5)
    at TCP.onStreamRead (node:internal/stream_base_commons:191:23)`

will try to figure it out.

Thanks!

**Reply by alxcohen** [reply]
btw, about starting like system service, is that any ideas why the same launch command had the crash by starting as a service?
result if running as a service 
`  --bind-pull, -p    The address to bind the ZeroMQ PULL endpoint to.
                                              [string] [default: "tcp://*:7113"]
  --name             An easily identifiable name for log output.
                                                  [string] [default: "mob-farm"]

Each option can be be overwritten with an environment variable by converting the
option to uppercase, replacing dashes with underscores and prefixing it with
`STF_TRIPROXY_` (e.g. `STF_TRIPROXY_BIND_PUB`).

Module did not self-register: '/root/.nvm/versions/node/v20.16.0/lib/node_modules/@devicefarmer/stf/node_modules/zeromq/build/Release/zmq.node'.`

**Comment by denis99999** [comment]
@alxcohen , perhaps you have also to value the `--ldap-search-field` argument? Regarding the system service take in mind it runs in root mode, not in user mode, so you have to ensure it can access to the required environment to launch the STF service, so in such case you should launch the STF service as a docker image instead that using the command line on your local file system

**Reply by alxcohen** [reply]
Nothing changes, same error(
Moved all to container.
ldapsearch passing ok 
Trying to use simple filter 
--auth-options '["--ldap-url","ldap://one.domain.com","--ldap-search-dn","DC=one,DC=domain,DC=com","--ldap-bind-dn","cn=stf_user,dc=one,dc=domain,dc=com","--ldap-bind-credentials","pass","--ldap-search-field","SamAccountName", "--ldap-search-filter","(&(SamAccountName={{username}}))"]'\
and still got the same
`stf        | 2024-08-13T14:31:04.668Z ERR/auth-ldap 55 [::ffff:127.0.0.1] Unexpected error InvalidCredentialsError: 80090308: LdapErr: DSID-0C090449, comment: AcceptSecurityContext error, data 52e, v3839
stf        |     at messageCallback (/app/node_modules/ldapjs/lib/client/client.js:1419:45)
stf        |     at Parser.onMessage (/app/node_modules/ldapjs/lib/client/client.js:1089:14)
stf        |     at Parser.emit (node:events:527:28)
stf        |     at Parser.write (/app/node_modules/ldapjs/lib/messages/parser.js:111:8)
stf        |     at Socket.onData (/app/node_modules/ldapjs/lib/client/client.js:1076:22)
stf        |     at Socket.emit (node:events:527:28)
stf        |     at addChunk (node:internal/streams/readable:324:12)
stf        |     at readableAddChunk (node:internal/streams/readable:297:9)
stf        |     at Socket.Readable.push (node:internal/streams/readable:234:10)
stf        |     at TCP.onStreamRead (node:internal/stream_base_commons:190:23)`

**Comment by denis99999** [comment]
@alxcohen , perhaps your issue is related to #743, take a look!
